fix: Handle Empty State for KeyValueEntry

jafar-albadarneh opened the pull request:
<!-- Please fill the entire template and make sure that all checkboxes are checked. -->

## Description

<!-- Explain the goal of this pull request by describing the issue it fixes or the need for the new or updated functionality. -->

Currently, `KeyValueEntry` handles empty arrays `[]` as  `null` state very well. 
The case happens when the model attribute is optional and its value is an empty array `[]`. 
More about the issue is described [here](https://github.com/filamentphp/filament/issues/11127) 

- [ * ] Visual changes (if any) are shown using screenshots/recordings of before and after.

***Before***
![image](https://github.com/filamentphp/filament/assets/21292175/ded7de2f-08e7-4b25-ac3b-cd30794933d7)


***After***
![image](https://github.com/filamentphp/filament/assets/21292175/ddd85754-f088-4a96-8a55-a8a91b438a90)


## Code style

<!-- Make sure code style follows the rest of the codebase. -->

- [x] `composer cs` command has been run.

## Testing

<!-- Ensure changes in this PR don't break existing functionality by testing it properly, either manually or by adding software tests. -->

- [X] Changes have been tested.

## Documentation

<!-- If new functionality has been added or existing functionality changed, please update the documentation. -->

- [X] Documentation is up-to-date.

jafar-albadarneh committed:
feat: Handle Empty State for KeyValueEntry

- Make sure to cast `null` to `[]`
- Update `key-value-entry.blade` to check for the empty state and render an empty state label
- Expose new method `emptyLabel()` to customize the placeholder text
- Add necessary defaults for `emptyLabel` attribute
- Update Docs

jafar-albadarneh committed:
fix: Handle Empty State for KeyValueEntry

- default to empty collection instead of `[]` for consistency

sandersjj commented:
[This bug report ](https://github.com/filamentphp/filament/issues/11128) might be releated.

danharrin commented on packages/infolists/src/Components/KeyValueEntry.php:
```diff
@@ -15,6 +16,23 @@ class KeyValueEntry extends Entry
 
+    protected string | Closure | null $emptyLabel = null;
+
+    public function getState(): mixed
+    {
+        $state = parent::getState();
+
+        if ($state === null) {
+            return collect();
+        }
+
+        if (is_array($state)) {
+            return collect($state);
+        }
+
+        return $state;
+    }
```
Could you please explain what this override is being used for? I think we should avoid it if possible

jafar-albadarneh replied:
@danharrin Of course, for some reason the state of the casted array attribute is being converted to `null`, thus causing the @foreach to fail within the template.
I tried my best to track where the value of the state being overwritten and set to `null` when it's `[]`
The proposed override attempts to change the state to `[]` when it's null.

You might have a good point about avoiding such an override since the template now checks for the state value before attempting the loop.
```php
@if($getState())
```

danharrin replied:
`@if (filled($state = ($getState() ?? [])))`, `@foreach ($state ...`

jafar-albadarneh replied:
@danharrin yeah sounds better. What's your take on the empty state template change?

danharrin replied:
What about it? My only comment would be renaming label to message, probably

danharrin commented:
What about it? My only comment would be renaming `label` to `message`, probably

jafar-albadarneh committed:
fix: Handle Empty State for KeyValueEntry

- remove state override method
- handle `null` state override from within the template
- update docs
- update i18 refs

danharrin commented on packages/infolists/docs/03-entries/06-key-value.md:
````diff
@@ -51,3 +51,14 @@ use Filament\Infolists\Components\KeyValueEntry;
     ->valueLabel('Property value')
 ```
+
+## Customizing the empty value pair's label
+
+You may customize the message for empty value pairs using the `emptyMessage()` method:
````
```suggestion
## Customizing the empty message

You may customize the message that is displayed when the entry is empty using the `emptyMessage()` method:
```

danharrin committed:
Update packages/infolists/docs/03-entries/06-key-value.md

danharrin commented:
Thank you for picking up on our code style and structure, and trying to make this fit into the rest of the framework as well as possible. It is appreciated!

jafar-albadarneh commented:
Makes complete sense
